### ISSUE TITLE
Report warnings right away, not at the end of the file

### ISSUE THREAD

**Issue by nad** [agda/agda#5306]
It might be easier to identify the cause of a performance regression in existing code if Agda notified the user about unsolved meta-variables as soon as possible (for both imported modules and the top-level one).

Related: #2514.

**Comment by nad** [comment]
@UlfNorell [suggested](https://github.com/agda/agda/issues/5314#issuecomment-817762767) that we should always report warnings right away.

**Comment by nad** [comment]
Highlighting for unsolved meta-variables and constraints is for some reason explicitly not generated right away:

https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/Interaction/Highlighting/Generate.hs#L90-L93

The commit implementing this behaviour (9a3a7043b5f14dfa58fd7b62c92de1445fccbc8a) seems to have been a reaction to a commit with the title "[ fix #3417 fix #4348 ] highlight warnings as they are created" (bb680eb7e0dd2606289d8bdc33ca916b1bde1a2f).

It seems to me as if highlighting for warnings can be generated twice:

* Here highlighting is generated for warnings other than `UnsolvedConstraints` and `UnsolvedMetaVariables`:

  https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/TypeChecking/Warnings.hs#L70-L73

* Here highlighting is generated for all warnings:

  https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/Interaction/Imports.hs#L1019-L1028

* Here highlighting is generated for `UnsolvedConstraints` and `UnsolvedMetaVariables`:

  https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/Interaction/Imports.hs#L1046

It might make sense to only generate the highlighting once.

**Comment by jespercockx** [comment]
@nad is this something you suggest we should fix before the release? If yes, are you planning to implement it soon?

**Comment by nad** [comment]
> @nad is this something you suggest we should fix before the release?

Yes. Regressions of the kind that @martinescardo reported could be easier to understand and fix with this change in place.

> If yes, are you planning to implement it soon?

I could take a look at it.

**Comment by nad** [comment]
@Saizan, I did not find any test cases related to your change discussed above (9a3a7043b5f14dfa58fd7b62c92de1445fccbc8a). Could you add a test case?

**Comment by nad** [comment]
I think that warnings for the following things are currently generated at the end of the module:
* Unsolved meta-variables corresponding to some goal.
* Other unsolved meta-variables.
* Unsolved constraints.

I suggest that we warn about unsolved meta-variables that do not correspond to a goal when they are frozen. I guess there is no need to warn about unsolved meta-variables corresponding to goals. One could perhaps warn about unsolved constraints that are blocked on frozen meta-variables, but there does not seem to be an asymptotically efficient way to find those constraints (the unsolved constraints are stored in lists).

Currently we print goal types together with the types of other unsolved meta-variables. If warnings are generated for the latter when they are frozen, and such warnings include enough information, then we could perhaps omit this information from the list of goals.

I also noticed that when the command-line interface is used, then warnings are not printed right away. Here is a test case:
```agda
{-# OPTIONS --warning=CoverageNoExactSplit #-}

open import Agda.Builtin.Bool
open import Agda.Builtin.Equality
open import Agda.Builtin.Nat

_∨_ : Bool → Bool → Bool
true  ∨ _     = true
_     ∨ true  = true
false ∨ false = false

fib : Nat → Nat
fib 0             = 0
fib 1             = 1
fib (suc (suc n)) = fib n + fib (suc n)

_ : fib 25 ≡ fib 24 + fib 23
_ = refl
```
The non-exact split in the definition of `_∨_` is highlighted more or less immediately in the Emacs mode, but it takes a while before the warning is printed when the command-line interface is used. The warning is currently printed twice:
```
Checking Test ([…]/Test.agda).
[…]/Test.agda:9,1-13
Exact splitting is enabled, but the following clause could not be
preserved as definitional equalities in the translation to a case
tree:
  _ ∨ true = true
when checking the definition of _∨_

———— All done; warnings encountered ————————————————————————

[…]/Test.agda:9,1-13
Exact splitting is enabled, but the following clause could not be
preserved as definitional equalities in the translation to a case
tree:
  _ ∨ true = true
when checking the definition of _∨_
```
The warning could be printed right away, and also together with all the warnings when everything is done.

**Comment by nad** [comment]
I started looking into this, but the code that handles warnings is for some reason rather complicated. For this reason I don't think we should change it shortly before the release of Agda 2.6.2.

Note also that highlighting unsolved meta-variables at the end of mutual blocks would not have made it easier to find the cause of [one of](https://github.com/agda/agda/issues/5279#issuecomment-801352579) the problems discussed elsewhere, because the unsolved meta-variable is in the same mutual block as the code that takes a lot of time/memory to type-check. However, emitting the "abstract definition without a type signature" warning right away could be of benefit to users porting code to Agda 2.6.2.
